File: scripts/fix-crosswalk.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function fixCrosswalk() {
    try {
        const csvPath = path.join(__dirname, '../crosswalks/NAICS_2022_to_NACE_Rev21_crosswalk.csv');
        
        console.log('Looking for CSV file at:', csvPath);
        
        if (!fs.existsSync(csvPath)) {
            console.error('CSV file not found:', csvPath);
            return;
        }

        console.log('🔧 FIXING NAICS-NACE CROSSWALK ISSUES...\n');

    const csvData = fs.readFileSync(csvPath, 'utf8');
    const lines = csvData.split('\n');
    const headers = lines[0].split(',').map(h => h.replace(/"/g, '').trim());
    
    const naceCodeIndex = headers.findIndex(h => h === 'NACE_Rev21_Code');
    const naicsCodeIndex = headers.findIndex(h => h === 'NAICS_2022_Code');
    const naceTitleIndex = headers.findIndex(h => h === 'NACE_Rev21_Title');
    const naicsTitleIndex = headers.findIndex(h => h === 'NAICS_2022_Title');
    
    let fixCount = 0;
    let issuesFound = [];
    
    // Keep header
    let fixedLines = [lines[0]];
    
    for (let i = 1; i < lines.length; i++) {
        const line = lines[i];
        if (!line.trim()) continue;
        
        const columns = line.split(',').map(col => col.replace(/"/g, '').trim());
        
        if (columns.length <= Math.max(naceCodeIndex, naicsCodeIndex)) {
            fixedLines.push(line);
            continue;
        }
        
        const naceCode = columns[naceCodeIndex];
        const naicsCode = columns[naicsCodeIndex];
        const naceTitle = columns[naceTitleIndex] || '';
        const naicsTitle = columns[naicsTitleIndex] || '';
        
        // Skip header rows and empty codes
        if (!naceCode || naceCode === 'NACE_Rev21_Code') {
            fixedLines.push(line);
            continue;
        }
        
        let shouldFix = false;
        let newNaceCode = naceCode;
        let newNaceTitle = naceTitle;
        let reason = '';
        
        // Fix 1: Agriculture mapped to tobacco manufacturing (weird crossover)
        if ((naicsCode === '111336' || naicsCode === '111998') && naceCode === '12') {
            // NAICS 111336: Fruit and Tree Nut Combination Farming should map to growing of fruits
            // NAICS 111998: All Other Miscellaneous Crop Farming should map to mixed farming
            if (naicsCode === '111336') {
                newNaceCode = '01.25'; // Growing of other tree and bush fruits and nuts
                newNaceTitle = 'Growing of other tree and bush fruits and nuts';
                reason = 'Fixed agriculture->tobacco crossover: Fruit farming to proper agriculture code';
            } else {
                newNaceCode = '01.50'; // Mixed farming  
                newNaceTitle = 'Mixed farming';
                reason = 'Fixed agriculture->tobacco crossover: Mixed crop farming to proper agriculture code';
            }
            shouldFix = true;
        }
        
        // Fix 2: Replace division-level codes (2-digit) with class-level codes (4-digit) where possible
        if (/^[0-9]{2}$/.test(naceCode) && naceCode !== '12') {
            // For most division codes, we can make them more specific
            const divisionCode = naceCode;
            switch(divisionCode) {
                case '01':
                    newNaceCode = '01.19'; // Growing of other non-perennial crops
                    newNaceTitle = 'Growing of other non-perennial crops';
                    reason = 'Made more specific: Division 01 -> Class 01.19';
                    shouldFix = true;
                    break;
                case '02':
                    newNaceCode = '02.10'; // Silviculture and other forestry activities  
                    newNaceTitle = 'Silviculture and other forestry activities';
                    reason = 'Made more specific: Division 02 -> Class 02.10';
                    shouldFix = true;
                    break;
                case '03':
                    newNaceCode = '03.11'; // Marine fishing
                    newNaceTitle = 'Marine fishing';
                    reason = 'Made more specific: Division 03 -> Class 03.11';
                    shouldFix = true;
                    break;
            }
        }
        
        // Apply fixes
        if (shouldFix) {
            // Update the NACE code and title columns
            const newColumns = [...columns];
            newColumns[naceCodeIndex] = newNaceCode;
            newColumns[naceTitleIndex] = newNaceTitle;
            
            const fixedLine = newColumns.map(col => `"${col}"`).join(',');
            fixedLines.push(fixedLine);
            
            fixCount++;
            issuesFound.push({
                line: i + 1,
                naics: naicsCode,
                naicsTitle: naicsTitle,
                oldNace: naceCode,
                newNace: newNaceCode,
                reason: reason
            });
            
            console.log(`Line ${i+1}: NAICS ${naicsCode} (${naicsTitle})`);
            console.log(`  OLD: ${naceCode} (${naceTitle})`); 
            console.log(`  NEW: ${newNaceCode} (${newNaceTitle})`);
            console.log(`  REASON: ${reason}\n`);
        } else {
            fixedLines.push(line);
        }
    }
    
    // Write fixed file
    const fixedContent = fixedLines.join('\n');
    fs.writeFileSync(csvPath, fixedContent);
    
    console.log('=== CROSSWALK FIXES SUMMARY ===');
    console.log(`✅ Fixed ${fixCount} problematic mappings`);
    console.log(`📁 Updated file: ${csvPath}`);
    
    if (fixCount > 0) {
        console.log('\n🔸 Issues Fixed:');
        issuesFound.forEach((issue, idx) => {
            console.log(`${idx+1}. ${issue.reason}`);
            console.log(`   NAICS ${issue.naics} -> OLD: ${issue.oldNace} -> NEW: ${issue.newNace}`);
        });
    }
    
    console.log('\n✨ Crosswalk fixes completed!');
    return { fixCount, issuesFound };
    
    } catch (error) {
        console.error('Error fixing crosswalk:', error);
        throw error;
    }
}

if (import.meta.url === `file://${process.argv[1]}`) {
    fixCrosswalk();
}

export { fixCrosswalk };

File: scripts/fix_crosswalk.py
#!/usr/bin/env python3
"""Fix NAICS-NACE crosswalk issues."""

import csv
import os
import sys

def fix_crosswalk():
    """Fix weird crossovers and use lowest-level NACE codes."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    csv_path = os.path.join(script_dir, '..', 'crosswalks', 'NAICS_2022_to_NACE_Rev21_crosswalk.csv')
    
    print(f"Looking for CSV file at: {csv_path}")
    
    if not os.path.exists(csv_path):
        print(f"ERROR: CSV file not found at {csv_path}")
        return
        
    print("FIXING NAICS-NACE CROSSWALK ISSUES...")
    
    fixes_made = []
    
    # Read and process the CSV
    rows = []
    with open(csv_path, 'r', encoding='utf-8', newline='') as f:
        reader = csv.reader(f)
        headers = next(reader)
        
        print(f"Headers: {headers}")
        
        # Find column indices
        nace_code_idx = headers.index('NACE_Rev21_Code')
        naics_code_idx = headers.index('NAICS_2022_Code') 
        nace_title_idx = headers.index('NACE_Rev21_Title')
        naics_title_idx = headers.index('NAICS_2022_Title')
        
        rows.append(headers)
        
        line_num = 2  # Start from line 2 (after header)
        for row in reader:
            if len(row) <= max(nace_code_idx, naics_code_idx):
                rows.append(row)
                continue
                
            naics_code = row[naics_code_idx].strip()
            nace_code = row[nace_code_idx].strip()
            naics_title = row[naics_title_idx].strip()
            nace_title = row[nace_title_idx].strip()
            
            should_fix = False
            new_nace_code = nace_code
            new_nace_title = nace_title
            reason = ""
            
            # Fix 1: Agriculture mapped to tobacco manufacturing (weird crossover)
            if naics_code == '111336' and nace_code == '12':
                # Fruit and Tree Nut Combination Farming should map to growing of fruits
                new_nace_code = '01.25'
                new_nace_title = 'Growing of other tree and bush fruits and nuts'
                reason = 'Fixed agriculture->tobacco crossover: Fruit farming to proper agriculture code'
                should_fix = True
                
            elif naics_code == '111998' and nace_code == '12':
                # All Other Miscellaneous Crop Farming should map to mixed farming
                new_nace_code = '01.50'
                new_nace_title = 'Mixed farming'
                reason = 'Fixed agriculture->tobacco crossover: Mixed crop farming to proper agriculture code'
                should_fix = True
                
            # Fix 2: Replace 2-digit division codes with 4-digit class codes where possible
            elif len(nace_code) == 2 and nace_code.isdigit() and nace_code != '12':
                if nace_code == '01':
                    new_nace_code = '01.19'
                    new_nace_title = 'Growing of other non-perennial crops'
                    reason = 'Made more specific: Division 01 -> Class 01.19'
                    should_fix = True
                elif nace_code == '02':
                    new_nace_code = '02.10' 
                    new_nace_title = 'Silviculture and other forestry activities'
                    reason = 'Made more specific: Division 02 -> Class 02.10'
                    should_fix = True
                elif nace_code == '03':
                    new_nace_code = '03.11'
                    new_nace_title = 'Marine fishing'
                    reason = 'Made more specific: Division 03 -> Class 03.11'
                    should_fix = True
                    
            if should_fix:
                # Apply the fix
                new_row = row.copy()
                new_row[nace_code_idx] = new_nace_code
                new_row[nace_title_idx] = new_nace_title
                rows.append(new_row)
                
                fix_info = {
                    'line': line_num,
                    'naics': naics_code,
                    'naics_title': naics_title,
                    'old_nace': nace_code,
                    'old_nace_title': nace_title,
                    'new_nace': new_nace_code,
                    'new_nace_title': new_nace_title,
                    'reason': reason
                }
                fixes_made.append(fix_info)
                
                print(f"Line {line_num}: NAICS {naics_code} ({naics_title})")
                print(f"  OLD: {nace_code} ({nace_title})")
                print(f"  NEW: {new_nace_code} ({new_nace_title})")
                print(f"  REASON: {reason}")
                print()
            else:
                rows.append(row)
                
            line_num += 1
    
    # Write the fixed CSV back
    with open(csv_path, 'w', encoding='utf-8', newline='') as f:
        writer = csv.writer(f)
        writer.writerows(rows)
    
    print("=== CROSSWALK FIXES SUMMARY ===")
    print(f"Fixed {len(fixes_made)} problematic mappings")
    print(f"Updated file: {csv_path}")
    
    if fixes_made:
        print("\\nIssues Fixed:")
        for i, fix in enumerate(fixes_made, 1):
            print(f"{i}. {fix['reason']}")
            print(f"   NAICS {fix['naics']} -> OLD: {fix['old_nace']} -> NEW: {fix['new_nace']}")
    
    print("\\nCrosswalk fixes completed!")
    return len(fixes_made)

if __name__ == '__main__':
    fix_crosswalk()